cmip6_downscaling/methods/bcsd.py

```diff
@@ -0,0 +1,249 @@
+import os
+
+from skdownscale.pointwise_models import BcAbsolute, BcRelative, PointWiseDownscaler
+
+from cmip6_downscaling.constants import ABSOLUTE_VARS, RELATIVE_VARS
+from cmip6_downscaling.data.cmip import load_cmip
+from cmip6_downscaling.data.observations import get_spatial_anomolies, load_obs
+from cmip6_downscaling.workflows.utils import (
+    load_paths,
+    rechunk_zarr_array,
+    regrid_dataset,
+    write_dataset,
+)
+
+connection_string = os.environ.get("AZURE_STORAGE_CONNECTION_STRING")
+# TODO: add templates for paths, maybe using prefect context. Do this once functions ported to prefect world in bcsd_flow.py
+
+
+def preprocess_bcsd(
+    gcm: str,
+    obs_id: str,
+    train_period_start: str,
+    train_period_end: str,
+    variable: str,
+    coarse_obs_path: str,
+    spatial_anomolies_path: str,
+    connection_string: str,
+    rerun: bool = True,
+) -> tuple[str, str]:
+    """Given GCM and selected obs dataset for a given variable and a given time period, at the coarse scale to match that
+    write out the coarsened version of the obs to match that GCM's grid and the spatial anomalies
+    for the obs over that time period associated with interpolating the coarse obs back to the original fine
+    obs resolution.
+
+    Parameters
+    ----------
+    gcm : str
+        Name of GCM
+    obs_id : str
+        Name of obs dataset (currently only supports 'ERA5')
+    train_period_start : str
+        Date for training period start (e.g. '1985')
+    train_period_end : str
+        Date for training period end (e.g. '2015')
+    variable : str
+        Variable of interest in CMIP conventions (e.g. 'tasmax')
+    coarse_obs_path : str
+        Path to write coarsened obs zarr dataset (e.g. 'az://bucket/out.zarr')
+    spatial_anomolies_path : str
+        Path to write coarsened obs zarr dataset (e.g. 'az://bucket/out.zarr')
+    connection_string : str
+        Connection string to give you read/write access to the out buckets specified above
+    rerun : bool, optional
+        [description], by default True
```

Comment:
note that we can remove this once we have caching implemented at the prefect level.